File: backend/core/ssdna_fjc.py
"""ssDNA freely-jointed chain (FJC) lookup.

Loads ``backend/data/ssdna_fjc_lookup.json`` (pre-computed by
``scripts/generate_ssdna_fjc_lookup.py``) once at module import time and
exposes per-length lookups for radius of gyration, end-to-end distance,
and the canonical bead positions.

Schema (post bin-based redesign)
--------------------------------
Each entry exposes:

  • ``r_ee_bin_edges_nm`` / ``rg_bin_edges_nm`` — HIST_BINS+1 floats.
  • ``bins`` — HIST_BINS items, each:
      { count, rep_r_ee_nm, rep_rg_nm, rep_positions, rg_subcounts }
    Non-empty bins carry ``rep_positions`` (the canonical-frame chain).
    ``rg_subcounts[k]`` is the count of samples in this R_ee bin whose
    Rg fell into Rg bin ``k`` — used by the frontend to re-compute the
    Rg distribution when the user crops the R_ee range.

Callers usually identify a configuration by ``bin_index`` ∈ [0, HIST_BINS-1].
``bin_positions(n_bp, bin_index)`` returns positions for that bin (or for
the nearest non-empty bin when ``bin_index`` falls in an empty cell).
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

import numpy as np

_LOOKUP_PATH = Path(__file__).resolve().parents[1] / "data" / "ssdna_fjc_lookup.json"


@lru_cache(maxsize=1)
def _load() -> dict:
    with _LOOKUP_PATH.open() as fh:
        return json.load(fh)


def dump_all() -> dict:
    """Return the raw lookup payload (metadata + entries).

    Used by the ``GET /api/ssdna-fjc-lookup`` endpoint so the frontend can
    fetch the table once and render ss-linker bridges in their natural
    FJC shape without round-tripping per-design.
    """
    return _load()


def metadata() -> dict:
    return dict(_load()["metadata"])


def supported_range() -> tuple[int, int]:
    md = metadata()
    lo, hi = md["bp_range"]
    return int(lo), int(hi)


def has_entry(n_bp: int) -> bool:
    return str(int(n_bp)) in _load()["entries"]


def entry(n_bp: int) -> dict:
    """Return the lookup entry for ``n_bp`` (fresh copy)."""
    if n_bp < 1:
        raise ValueError(f"n_bp must be >= 1, got {n_bp}")
    key = str(int(n_bp))
    entries = _load()["entries"]
    if key not in entries:
        lo, hi = supported_range()
        raise ValueError(
            f"ssDNA FJC lookup has no entry for n_bp={n_bp}; "
            f"table covers {lo}..{hi}. Regenerate "
            "scripts/generate_ssdna_fjc_lookup.py to extend the range."
        )
    return dict(entries[key])


def num_bins(n_bp: int) -> int:
    return len(entry(n_bp)["bins"])


def _resolve_bin(n_bp: int, bin_index: int) -> tuple[int, dict]:
    """Return (resolved_index, bin_payload) for the nearest non-empty bin
    to ``bin_index``. Raises if no bin in the entry is occupied."""
    bins = entry(n_bp)["bins"]
    if not bins:
        raise ValueError(f"FJC entry n_bp={n_bp} has no bins")
    n = len(bins)
    raw = int(bin_index) % n
    if bins[raw]["count"] > 0 and bins[raw].get("rep_positions") is not None:
        return raw, bins[raw]
    # Walk outward until we find a populated bin.
    for d in range(1, n + 1):
        for cand in (raw - d, raw + d):
            if 0 <= cand < n and bins[cand]["count"] > 0 and bins[cand].get("rep_positions") is not None:
                return cand, bins[cand]
    raise ValueError(f"FJC entry n_bp={n_bp} has no occupied bins")


def bin_positions(n_bp: int, bin_index: int) -> np.ndarray:
    """Return ``(n_bp, 3)`` canonical positions for the chosen bin (or the
    nearest non-empty bin)."""
    _idx, b = _resolve_bin(n_bp, bin_index)
    pos = np.asarray(b["rep_positions"], dtype=float)
    if pos.ndim != 2 or pos.shape[1] != 3:
        raise ValueError(f"FJC bin n_bp={n_bp}, bin={bin_index} has malformed positions {pos.shape}")
    return pos


def bin_r_ee(n_bp: int, bin_index: int) -> float:
    _idx, b = _resolve_bin(n_bp, bin_index)
    return float(b["rep_r_ee_nm"])


def bin_rg(n_bp: int, bin_index: int) -> float:
    _idx, b = _resolve_bin(n_bp, bin_index)
    return float(b["rep_rg_nm"])


def resolve_bin_index(n_bp: int, bin_index: int) -> int:
    """Resolve an arbitrary ``bin_index`` to the actual non-empty bin used
    by the lookup (skipping empties via nearest-non-empty walk)."""
    idx, _ = _resolve_bin(n_bp, bin_index)
    return idx


def default_bin_index(n_bp: int) -> int:
    """Bin whose midpoint R_ee is closest to the ensemble mean. Used as the
    initial selection for a freshly relaxed linker."""
    e = entry(n_bp)
    mean = float(e.get("r_ee_mean_nm", 0.0))
    edges = e["r_ee_bin_edges_nm"]
    midpoints = [0.5 * (edges[k] + edges[k + 1]) for k in range(len(edges) - 1)]
    if not midpoints:
        return 0
    target_idx = int(np.argmin([abs(m - mean) for m in midpoints]))
    idx, _ = _resolve_bin(n_bp, target_idx)
    return idx


def transform_to_chord(
    positions: np.ndarray,
    anchor_a: np.ndarray,
    anchor_b: np.ndarray,
) -> np.ndarray:
    """Anisotropically stretch + rotate + translate canonical positions onto
    the live chord A→B. ``positions[0]`` lands at ``anchor_a``; the last
    bead lies along the chord at the canonical R_ee distance.
    """
    a = np.asarray(anchor_a, dtype=float)
    b = np.asarray(anchor_b, dtype=float)
    chord = b - a
    cl = float(np.linalg.norm(chord))
    if cl < 1e-9:
        return positions + a

    target = chord / cl
    src = np.array([1.0, 0.0, 0.0])
    s = float(np.linalg.norm(np.cross(src, target)))
    c = float(np.dot(src, target))
    if s < 1e-12:
        if c > 0:
            return positions + a
        rot = np.diag([-1.0, -1.0, 1.0])
        return positions @ rot.T + a

    axis = np.cross(src, target) / s
    K = np.array(
        [
            [0.0, -axis[2], axis[1]],
            [axis[2], 0.0, -axis[0]],
            [-axis[1], axis[0], 0.0],
        ]
    )
    rot = np.eye(3) + s * K + (1 - c) * (K @ K)
    return positions @ rot.T + a

File: frontend/src/scene/ssdna_fjc.js
/**
 * ssDNA freely-jointed chain lookup — client-side cache + chord-fit transform.
 *
 * Schema (bin-based redesign): each lookup entry exposes an R_ee histogram
 * with `hist_bins` bins. Each non-empty bin carries one representative
 * shape (`rep_positions`) and per-Rg-bin counts (`rg_subcounts`) so the
 * linker-config modal can re-compute the Rg distribution under any R_ee
 * range crop without round-tripping the backend.
 *
 * Used by `overhang_link_arcs.js` to render the chosen bridge shape and
 * by `linker_config_modal.js` for the interactive shape picker.
 */

import * as THREE from 'three'

let _lookup = null
let _loadPromise = null
let _onLoaded = []

export function ensureLoaded() {
  if (_lookup) return Promise.resolve(_lookup)
  if (_loadPromise) return _loadPromise
  _loadPromise = fetch('/api/ssdna-fjc-lookup', {
    headers: { 'Accept': 'application/json' },
    cache: 'no-store',
  })
    .then(r => {
      if (!r.ok) throw new Error(`ssdna-fjc-lookup HTTP ${r.status}`)
      return r.json()
    })
    .then(payload => {
      _lookup = payload
      const cfgCount = payload?.metadata?.hist_bins
      if (cfgCount !== 40) {
        console.warn('[ssdna_fjc] unexpected hist_bins:', cfgCount,
          '(metadata:', payload?.metadata, ')')
      }
      const subs = _onLoaded.splice(0)
      for (const fn of subs) {
        try { fn(payload) } catch (e) { console.error('[ssdna_fjc] onLoaded subscriber threw:', e) }
      }
      return payload
    })
    .catch(err => {
      console.warn('[ssdna_fjc] lookup load failed; ss bridges will fall back to Bezier:', err)
      _loadPromise = null
      throw err
    })
  return _loadPromise
}

export function isLoaded() { return _lookup != null }

export function onLoaded(fn) {
  if (_lookup) { fn(_lookup); return () => {} }
  _onLoaded.push(fn)
  return () => { _onLoaded = _onLoaded.filter(f => f !== fn) }
}

export function metadata() { return _lookup?.metadata ?? null }

export function lookupEntry(nBp) {
  if (!_lookup) return null
  const key = String(Math.max(1, Math.round(nBp)))
  return _lookup.entries?.[key] ?? null
}

/** Convert an arbitrary bin index to the nearest occupied bin (with a rep
 *  shape). Returns null if the entry has no occupied bins. */
export function resolveBinIndex(nBp, binIndex) {
  const entry = lookupEntry(nBp)
  if (!entry) return null
  const bins = entry.bins ?? []
  if (!bins.length) return null
  const n = bins.length
  let raw = ((binIndex % n) + n) % n
  if (bins[raw]?.count > 0 && bins[raw]?.rep_positions) return raw
  for (let d = 1; d <= n; d++) {
    for (const cand of [raw - d, raw + d]) {
      if (cand >= 0 && cand < n && bins[cand]?.count > 0 && bins[cand]?.rep_positions) {
        return cand
      }
    }
  }
  return null
}

/** Bin whose midpoint R_ee is closest to the ensemble mean. */
export function defaultBinIndex(nBp) {
  const entry = lookupEntry(nBp)
  if (!entry) return null
  const mean = entry.r_ee_mean_nm ?? 0
  const edges = entry.r_ee_bin_edges_nm ?? []
  let best = -1
  let bestErr = Infinity
  for (let k = 0; k < edges.length - 1; k++) {
    const mid = 0.5 * (edges[k] + edges[k + 1])
    const err = Math.abs(mid - mean)
    if (err < bestErr) { bestErr = err; best = k }
  }
  return resolveBinIndex(nBp, best)
}

export function binPositions(nBp, binIndex) {
  const idx = resolveBinIndex(nBp, binIndex)
  if (idx == null) return null
  return lookupEntry(nBp).bins[idx].rep_positions
}

export function binRee(nBp, binIndex) {
  const idx = resolveBinIndex(nBp, binIndex)
  if (idx == null) return null
  return lookupEntry(nBp).bins[idx].rep_r_ee_nm
}

export function binRg(nBp, binIndex) {
  const idx = resolveBinIndex(nBp, binIndex)
  if (idx == null) return null
  return lookupEntry(nBp).bins[idx].rep_rg_nm
}

/** Return the Rg histogram counts for samples whose R_ee falls inside
 *  [rEeMin, rEeMax] — by summing `rg_subcounts` of bins in the range. */
export function filteredRgHistogram(nBp, rEeMin, rEeMax) {
  const entry = lookupEntry(nBp)
  if (!entry) return null
  const edges = entry.r_ee_bin_edges_nm ?? []
  const bins = entry.bins ?? []
  const rgEdges = entry.rg_bin_edges_nm ?? []
  const nRg = rgEdges.length - 1
  if (nRg <= 0) return null
  const out = new Array(nRg).fill(0)
  for (let k = 0; k < bins.length; k++) {
    const lo = edges[k], hi = edges[k + 1]
    if (hi < rEeMin - 1e-9 || lo > rEeMax + 1e-9) continue
    const sub = bins[k]?.rg_subcounts ?? []
    for (let j = 0; j < nRg; j++) {
      out[j] += sub[j] ?? 0
    }
  }
  return { bin_edges_nm: rgEdges, counts: out }
}

/**
 * Anisotropic stretch + rotate + translate canonical positions onto the
 * live anchor chord A→B. y, z preserved; x scaled so the last bead lands
 * at chord_length along the chord direction.
 */
export function transformToChord(canonical, anchorA, anchorB, rEe) {
  const chord = new THREE.Vector3().subVectors(anchorB, anchorA)
  const chordLen = chord.length()
  if (rEe < 1e-9 || chordLen < 1e-9) {
    return canonical.map(() => anchorA.clone())
  }
  const k = chordLen / rEe
  const target = chord.clone().normalize()
  const src = new THREE.Vector3(1, 0, 0)

  const axis = new THREE.Vector3().crossVectors(src, target)
  const sin = axis.length()
  const cos = src.dot(target)
  let rot
  if (sin < 1e-12) {
    rot = cos > 0
      ? new THREE.Matrix3().identity()
      : new THREE.Matrix3().set(-1, 0, 0,  0, -1, 0,  0, 0, 1)
  } else {
    axis.divideScalar(sin)
    const x = axis.x, y = axis.y, z = axis.z
    const C = 1 - cos
    rot = new THREE.Matrix3().set(
      cos + x*x*C,     x*y*C - z*sin,  x*z*C + y*sin,
      y*x*C + z*sin,   cos + y*y*C,    y*z*C - x*sin,
      z*x*C - y*sin,   z*y*C + x*sin,  cos + z*z*C,
    )
  }

  const out = new Array(canonical.length)
  const tmp = new THREE.Vector3()
  for (let i = 0; i < canonical.length; i++) {
    const [cx, cy, cz] = canonical[i]
    tmp.set(cx * k, cy, cz).applyMatrix3(rot).add(anchorA)
    out[i] = tmp.clone()
  }
  return out
}

/** Convenience: world-frame chain for a given (n_bp, binIndex) between anchors. */
export function fjcChainBetween(nBp, anchorA, anchorB, binIndex = null) {
  const idx = binIndex == null ? defaultBinIndex(nBp) : resolveBinIndex(nBp, binIndex)
  if (idx == null) return null
  const positions = binPositions(nBp, idx)
  const rEe = binRee(nBp, idx)
  if (!positions || rEe == null) return null
  return transformToChord(positions, anchorA, anchorB, rEe)
}
